packages/vite/src/node/server/searchRoot.ts

```diff
@@ -0,0 +1,48 @@
+import fs from 'fs'
+import { dirname } from 'path'
+import { join } from 'path'
+
+// https://github.com/vitejs/vite/issues/2820#issuecomment-812495079
+const ROOT_FILES = [
+  '.git',
+
+  // https://pnpm.js.org/workspaces/
+  'pnpm-workspace.yaml'
+
+  // https://rushjs.io/pages/advanced/config_files/
+  // 'rush.json',
+
+  // https://nx.dev/latest/react/getting-started/nx-setup
+  // 'workspace.json',
+  // 'nx.json'
+]
+
+// npm: https://docs.npmjs.com/cli/v7/using-npm/workspaces#installing-workspaces
+// yarn: https://classic.yarnpkg.com/en/docs/workspaces/#toc-how-to-use-it
+function hasWorkspacePackageJSON(root: string): boolean {
+  const path = join(root, 'package.json')
+  if (!fs.existsSync(path)) return false
+  const content = JSON.parse(fs.readFileSync(path, 'utf-8')) || {}
+  return !!content.workspaces
+}
+
+function hasRootFile(root: string): boolean {
+  for (const file of ROOT_FILES) {
+    if (fs.existsSync(join(root, file))) return true
+  }
+  return false
```

Comment:
Maybe we can bring this into a one-liner that may be better readable 🤔 

```suggestion
  ROOT_FILES.some((file) => fs.existsSync(join(root, file)))
```